File: native/android/app/src/main/java/com/mediconnect/nativeapp/PatientSettingsEditor.kt
package com.mediconnect.nativeapp

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.Instant
import kotlin.coroutines.coroutineContext

enum class SettingsStep { CLOSED, LOADING, EDITING, SAVING, SAVED, LOAD_FAILED, SAVE_UNCERTAIN, DENIED, MISSING }
data class SettingsState(val step: SettingsStep = SettingsStep.CLOSED, val snapshot: PatientSettingsSnapshot? = null,
                         val draft: PatientSettingsDraft? = null, val invalid: Boolean = false) {
    val busy get() = step == SettingsStep.LOADING || step == SettingsStep.SAVING
    val editable get() = step == SettingsStep.EDITING || step == SettingsStep.SAVED
}

class PatientSettingsEditor(private val service: PatientSettingsService, private val maxNameLength: Int,
                            private val scope: CoroutineScope) {
    private val mutable = MutableStateFlow(SettingsState())
    val state = mutable.asStateFlow()
    private var identity: Identity? = null
    private var operation: Job? = null
    private var generation = 0
    fun close() { generation++; operation?.cancel(); operation = null; identity = null; mutable.value = SettingsState() }
    fun open(value: Identity) {
        close()
        if (value.role != Role.PATIENT || value.expiresAt <= Instant.now()) return
        identity = value; reload()
    }
    fun edit(draft: PatientSettingsDraft) {
        if (mutable.value.editable) mutable.value = mutable.value.copy(step = SettingsStep.EDITING,
            draft = draft.copy(preferences = draft.preferences?.toMap()), invalid = false)
    }
    fun discard() {
        val current = mutable.value
        if (current.editable && current.snapshot != null) mutable.value = current.copy(step = SettingsStep.EDITING,
            draft = PatientSettingsDraft(current.snapshot), invalid = false)
    }
    fun reload() {
        val owner = identity ?: return
        if (mutable.value.busy) return
        val current = ++generation
        mutable.value = SettingsState(SettingsStep.LOADING)
        operation = scope.launch {
            try {
                val snapshot = service.load(owner)
                coroutineContext.ensureActive()
                if (current != generation) return@launch
                require(snapshot.subject == owner.subject)
                mutable.value = SettingsState(SettingsStep.EDITING, snapshot, PatientSettingsDraft(snapshot))
            } catch (cancelled: CancellationException) { throw cancelled }
            catch (failure: Exception) { if (current == generation) mutable.value = SettingsState(readFailure(failure)) }
        }
    }
    fun save() {
        val owner = identity ?: return
        val before = mutable.value
        val snapshot = before.snapshot ?: return
        val draft = before.draft ?: return
        if (!before.editable) return
        val changes = try { PatientSettingsData.changes(snapshot, draft, maxNameLength) }
        catch (_: Exception) { mutable.value = before.copy(invalid = true); return }
        if (changes.length() == 0) return
        val current = ++generation
        mutable.value = before.copy(step = SettingsStep.SAVING, invalid = false)
        operation = scope.launch {
            try {
                service.save(owner, snapshot, draft)
                coroutineContext.ensureActive()
                if (current != generation) return@launch
                val saved = service.load(owner)
                coroutineContext.ensureActive()
                if (current != generation) return@launch
                require(saved.subject == owner.subject)
                require(draft.name == snapshot.name || saved.name == draft.name)
                require(draft.phone == snapshot.phone || saved.phone == draft.phone)
                require(draft.address == snapshot.address || saved.address == draft.address)
                require(draft.preferences == snapshot.preferences || saved.preferences == draft.preferences)
                mutable.value = SettingsState(SettingsStep.SAVED, saved, PatientSettingsDraft(saved))
            } catch (cancelled: CancellationException) { throw cancelled }
            catch (failure: Exception) {
                if (current == generation) mutable.value = if (failure is ApiFailure && failure.status in setOf(401, 403))
                    SettingsState(SettingsStep.DENIED) else before.copy(step = SettingsStep.SAVE_UNCERTAIN)
            }
        }
    }
    private fun readFailure(failure: Exception) = when ((failure as? ApiFailure)?.status) {
        401, 403 -> SettingsStep.DENIED
        404 -> SettingsStep.MISSING
        else -> SettingsStep.LOAD_FAILED
    }
}

File: native/ios/Sources/PatientSettingsEditor.swift
import Combine
import Foundation

enum SettingsStep: Equatable { case closed, loading, editing, saving, saved, loadFailed, saveUncertain, denied, missing }
struct SettingsState: Equatable {
    var step: SettingsStep = .closed
    var snapshot: PatientSettingsSnapshot?
    var draft: PatientSettingsDraft?
    var invalid = false
    var busy: Bool { step == .loading || step == .saving }
    var editable: Bool { step == .editing || step == .saved }
}

@MainActor final class PatientSettingsEditor: ObservableObject {
    @Published private(set) var state = SettingsState()
    private let service: PatientSettingsService
    private let maxNameLength: Int
    private var identity: Identity?
    private(set) var operation: Task<Void, Never>?
    private var generation = 0
    init(service: PatientSettingsService, maxNameLength: Int) {
        self.service = service; self.maxNameLength = maxNameLength
    }
    func close() { generation += 1; operation?.cancel(); operation = nil; identity = nil; state = SettingsState() }
    func open(_ value: Identity) {
        close()
        guard value.role == .patient, value.expiresAt > Date() else { return }
        identity = value; reload()
    }
    func edit(_ draft: PatientSettingsDraft) {
        guard state.editable, state.draft != draft else { return }
        state.step = .editing; state.draft = draft; state.invalid = false
    }
    func discard() {
        guard state.editable, let snapshot = state.snapshot else { return }
        state.step = .editing; state.draft = PatientSettingsDraft(snapshot: snapshot); state.invalid = false
    }
    func reload() {
        guard let identity, !state.busy else { return }
        generation += 1; let current = generation
        state = SettingsState(step: .loading)
        operation = Task { [weak self] in
            guard let self else { return }
            do {
                let snapshot = try await service.load(identity: identity)
                try Task.checkCancellation()
                guard generation == current else { return }
                guard snapshot.subject == identity.subject else { throw MobileFailure.response }
                state = SettingsState(step: .editing, snapshot: snapshot, draft: PatientSettingsDraft(snapshot: snapshot))
            } catch is CancellationError { return }
            catch {
                guard generation == current, !Task.isCancelled else { return }
                let status = (error as? HTTPFailure)?.status
                state = SettingsState(step: status == 401 || status == 403 ? .denied : (status == 404 ? .missing : .loadFailed))
            }
        }
    }
    func save() {
        guard let identity, state.editable, let snapshot = state.snapshot, let draft = state.draft else { return }
        let changes: [String: Any]
        do { changes = try PatientSettingsData.changes(snapshot: snapshot, draft: draft, maxNameLength: maxNameLength) }
        catch { state.invalid = true; return }
        guard !changes.isEmpty else { return }
        let before = state
        generation += 1; let current = generation
        state.step = .saving; state.invalid = false
        operation = Task { [weak self] in
            guard let self else { return }
            do {
                try await service.save(identity: identity, snapshot: snapshot, draft: draft)
                try Task.checkCancellation()
                guard generation == current else { return }
                let saved = try await service.load(identity: identity)
                try Task.checkCancellation()
                guard generation == current else { return }
                guard saved.subject == identity.subject,
                      draft.name == snapshot.name || saved.name == draft.name,
                      draft.phone == snapshot.phone || saved.phone == draft.phone,
                      draft.address == snapshot.address || saved.address == draft.address,
                      draft.preferences == snapshot.preferences || saved.preferences == draft.preferences else { throw MobileFailure.response }
                state = SettingsState(step: .saved, snapshot: saved, draft: PatientSettingsDraft(snapshot: saved))
            } catch is CancellationError { return }
            catch {
                guard generation == current, !Task.isCancelled else { return }
                let status = (error as? HTTPFailure)?.status
                if status == 401 || status == 403 { state = SettingsState(step: .denied) }
                else { state = before; state.step = .saveUncertain }
            }
        }
    }
}
